In [2]:
import pandas as pd

In [3]:
from keras.models import Sequential

In [4]:
from keras.layers import Dense

In [8]:
##Read the dataset

In [9]:
data = pd.read_csv("diabetes.csv")

In [10]:
data.head()

Pregnancies  Glucose  BloodPressure  ...  DiabetesPedigreeFunction  Age  Outcome
0            6      148             72  ...                     0.627   50        1
1            1       85             66  ...                     0.351   31        0
2            8      183             64  ...                     0.672   32        1
3            1       89             66  ...                     0.167   21        0
4            0      137             40  ...                     2.288   33        1

[5 rows x 9 columns]

In [16]:
data.describe()

Pregnancies     Glucose  ...         Age     Outcome
count   768.000000  768.000000  ...  768.000000  768.000000
mean      3.845052  120.894531  ...   33.240885    0.348958
std       3.369578   31.972618  ...   11.760232    0.476951
min       0.000000    0.000000  ...   21.000000    0.000000
25%       1.000000   99.000000  ...   24.000000    0.000000
50%       3.000000  117.000000  ...   29.000000    0.000000
75%       6.000000  140.250000  ...   41.000000    1.000000
max      17.000000  199.000000  ...   81.000000    1.000000

[8 rows x 9 columns]

The data has total of 768 rows and 9 columns


In [11]:
##checking any missing values

In [12]:
data.isnull().sum(axis=0)

Pregnancies                 0
Glucose                     0
BloodPressure               0
SkinThickness               0
Insulin                     0
BMI                         0
DiabetesPedigreeFunction    0
Age                         0
Outcome                     0
dtype: int64

There are no missing values in the data

In [13]:
## spliting the data into X and Y
X = data.iloc[:,:-1]
Y = data.iloc[:,-1] 

In [15]:
print(X)
print(Y)

     Pregnancies  Glucose  BloodPressure  ...   BMI  DiabetesPedigreeFunction  Age
0              6      148             72  ...  33.6                     0.627   50
1              1       85             66  ...  26.6                     0.351   31
2              8      183             64  ...  23.3                     0.672   32
3              1       89             66  ...  28.1                     0.167   21
4              0      137             40  ...  43.1                     2.288   33
..           ...      ...            ...  ...   ...                       ...  ...
763           10      101             76  ...  32.9                     0.171   63
764            2      122             70  ...  36.8                     0.340   27
765            5      121             72  ...  26.2                     0.245   30
766            1      126             60  ...  30.1                     0.349   47
767            1       93             70  ...  30.4                     0.315   23

[76

Spliting the dataset into train data and test data by importint train_test_split from Scikit-learn

In [18]:
from sklearn.model_selection import train_test_split

x_train,x_test,y_train,y_test = train_test_split(X,Y,test_size = 0.2, random_state = 1234, stratify = Y)

The dataset has been splitted in 80%-20% ratio

In [20]:
x_train

Pregnancies  Glucose  BloodPressure  ...   BMI  DiabetesPedigreeFunction  Age
54             7      150             66  ...  34.7                     0.718   42
159           17      163             72  ...  40.9                     0.817   47
30             5      109             75  ...  36.0                     0.546   60
622            6      183             94  ...  40.8                     1.461   45
462            8       74             70  ...  35.3                     0.705   39
..           ...      ...            ...  ...   ...                       ...  ...
84             5      137            108  ...  48.8                     0.227   37
223            7      142             60  ...  28.8                     0.687   61
36            11      138             76  ...  33.2                     0.420   35
457            5       86             68  ...  30.2                     0.364   24
242            3      139             54  ...  25.6                     0.402   22

[614 rows x 8 columns]

In [21]:
x_test

Pregnancies  Glucose  BloodPressure  ...   BMI  DiabetesPedigreeFunction  Age
511            0      139             62  ...  22.1                     0.207   21
707            2      127             46  ...  34.4                     0.176   22
596            0       67             76  ...  45.3                     0.194   46
595            0      188             82  ...  32.0                     0.682   22
283            7      161             86  ...  30.4                     0.165   47
..           ...      ...            ...  ...   ...                       ...  ...
385            1      119             54  ...  22.3                     0.205   24
383            1       90             62  ...  25.1                     1.268   25
481            0      123             88  ...  35.2                     0.197   29
136            0      100             70  ...  30.8                     0.597   21
394            4      158             78  ...  32.9                     0.803   31

[154 rows x 8 columns]

Defining the keras model

In [24]:
model = Sequential()

model.add(Dense(24,
                input_shape = (8,),
                activation = "relu",
                kernel_initializer = "RandomNormal"))

model.add(Dense(12,
                activation = "relu",
                kernel_initializer = "RandomNormal"))

model.add(Dense(1,
                activation="sigmoid"))

For this model sequential model has been selected, Relu has been selected for activation function and randomnormal for initializer

In [25]:
##compiling the model
model.compile(optimizer = "adam",
              loss = "binary_crossentropy",
              metrics = ["accuracy"])

The optimizer chosen was Adam and for loss function Binary cross entropy has been selected

Now run the model with epochs 160 and batch size of 10

In [26]:
model.fit(x_train,y_train, epochs = 160,batch_size = 10)

Epoch 1/160
62/62 [==============================] - 1s 1ms/step - loss: 0.8292 - accuracy: 0.5631
Epoch 2/160
62/62 [==============================] - 0s 1ms/step - loss: 0.6520 - accuracy: 0.6478
Epoch 3/160
62/62 [==============================] - 0s 1ms/step - loss: 0.6450 - accuracy: 0.6302
Epoch 4/160
62/62 [==============================] - 0s 1ms/step - loss: 0.6216 - accuracy: 0.6336
Epoch 5/160
62/62 [==============================] - 0s 1ms/step - loss: 0.6295 - accuracy: 0.6577
Epoch 6/160
62/62 [==============================] - 0s 1ms/step - loss: 0.6001 - accuracy: 0.6829
Epoch 7/160
62/62 [==============================] - 0s 1ms/step - loss: 0.6172 - accuracy: 0.6775
Epoch 8/160
62/62 [==============================] - 0s 1ms/step - loss: 0.6075 - accuracy: 0.6684
Epoch 9/160
62/62 [==============================] - 0s 1ms/step - loss: 0.5914 - accuracy: 0.7066
Epoch 10/160
62/62 [==============================] - 0s 1ms/step - loss: 0.5988 - accuracy: 0.6781
Epoch 11/

Now checking the accuracy

In [27]:
accuracy_test = model.evaluate(x_test,y_test)

5/5 [==============================] - 0s 2ms/step - loss: 0.5820 - accuracy: 0.7273


In [28]:
#get predicted value and probability of Y
Y_predict = model.predict_classes(x_test)

y_pred_prob = model.predict(x_test)

/usr/local/lib/python3.6/dist-packages/tensorflow/python/keras/engine/sequential.py:450: UserWarning: `model.predict_classes()` is deprecated and will be removed after 2021-01-01. Please use instead:* `np.argmax(model.predict(x), axis=-1)`,   if your model does multi-class classification   (e.g. if it uses a `softmax` last-layer activation).* `(model.predict(x) > 0.5).astype("int32")`,   if your model does binary classification   (e.g. if it uses a `sigmoid` last-layer activation).
  warnings.warn('`model.predict_classes()` is deprecated and '


In [29]:
#build confusion matrix
from sklearn.metrics import confusion_matrix, classification_report
cm = confusion_matrix(y_test, Y_predict)
cr = classification_report(y_test,Y_predict)

In [30]:
print(accuracy_test)
print(cm)

[0.5819634795188904, 0.7272727489471436]
[[90 10]
 [32 22]]


The model as predicted the test data with 72.7% accuracy that is it has predicted 112 records correctly out of 154 record